dynamic_array/include/dynamic_array/array.h

```diff
@@ -0,0 +1,303 @@
+#pragma once
+
+#include <iostream>
+
+namespace td {
+
+constexpr int min_capacity = 16;
+// https://stackoverflow.com/questions/1100311/what-is-the-ideal-growth-rate-for-a-dynamically-allocated-array
+constexpr int growth_factor = 2;
+constexpr int shrink_factor = 4;
+
+constexpr int index_not_found = -1;
+
+template <typename T>
+class Array;
+
+template <typename T>
+void swap(Array<T>& rhs, Array<T>& lhs);
+
+// A dynamic array template
+template <typename T>
+class Array {
+ public:
+  Array();
+  Array(size_t capacity);
+  Array(std::initializer_list<T>&& il);
+
+  Array(const Array<T>& other);
+  Array(Array<T>&& other);
+
+  Array<T>& operator=(const Array<T>& other);
+  Array<T>& operator=(Array<T>&& other);
+
+  ~Array();
+
+  // Return number of items are currently stored in array.
+  size_t size();
+
+  // Return number of items it can hold.
+  size_t capacity();
+
+  // Return array is empty or not.
+  bool is_empty();
+
+  // Return items at |index|.
+  T item_at(int index);
+
+  // Append |item| to the end of array.
+  void append(const T& item);
+
+  // Insert |item| at |index|.
+  void insert(const T& item, int index);
+
+  // Prepend |item| to the array.
+  void prepend(const T& item);
+
+  // Remove last item and return it.
+  T pop();
+
+  // Remove item at |index|
+  void remove_at(int index);
+
+  // Look for |item|, remove index holding it.
+  void remove(const T& item);
+
+  // Look for |item|, return first index with this
+  // value, returns |kIndexNotFound| if not found
+  int find(const T& item);
+
+  // Swap values inside |first| and |second|.
+  // Follow copy-and-swap idiom
+  // https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom
+  friend void swap<T>(Array<T>& rhs, Array<T>& lsh);
+
+ private:
+  // If |index| is a negative value or greater than
+  // |size_|, log an error and terminate program.
+  void throw_out_of_range_error_if_needed(int index);
```

Comment:
Let's use `std::exception` family. Or you want to create your custom `exception` class inherit from `std::exception`?